packages/peregrine/src/Peregrine/peregrine.js

```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import { arrayOf, oneOf } from 'prop-types';
+import { ToastContextProvider } from '../Toasts';
+import { WindowSizeContextProvider } from '../hooks/useWindowSize';
+
+const contextProviders = {
+    TOASTS: ToastContextProvider,
+    WINDOW_SIZE: WindowSizeContextProvider
+};
+
+const Peregrine = ({ children, providers }) => {
+    return providers.reduceRight((memo, providerKey) => {
+        const ContextProvider = contextProviders[providerKey];
+        return <ContextProvider>{memo}</ContextProvider>;
+    }, children);
+};
+
+Peregrine.propTypes = {
+    providers: arrayOf(oneOf(Object.keys(contextProviders)))
```

Comment:
There a couple of problems with this approach.

1. Opting out of a provider is difficult, because you have to pass the rest of the entire array as a prop. Once we have a dozen providers, that won't really work.
1. Since all of the providers are statically imported, you're not saving anything by opting out of a provider. The same amount of code is bundled and served, even if you pass an empty array. If there's no bundle savings by opting out, no one should ever opt out of a provider.

Reply:
@jimbo are you suggesting that people pass in a list of the Providers they wish to use? So they'd import in `App` for example and pass the list?

Also can we use dynamic import/suspense here?

Reply:
@jimbo - we can switch to using a `disabledProviders` array prop, but in the current state wouldn't solve the bundle size concern. Are you suggesting we explore lazy loading of these components, or just stripping config out entirely and not giving users the ability to configure providers?

Reply:
We definitely want users to be able to configure which providers are imported/used. Can you explore lazy loading the providers based on a flag?

Reply:
@sirugh I don't think we can state that quite so unequivocally.

### Case 0: no unified provider

The default use case here is that we don't have a unified context provider. In this case, an app imports all of the context providers it needs individually.

- 👍 Only the providers needed are imported
- 👍 Only the providers needed are rendered
- 👍 Providers are imported & rendered synchronously
- 👎 It's verbose

### Case 1: unified provider, no config

The first proposal, which @tjwiebell originally created in this PR, is a simple unified provider that imports all of the common providers and renders them.

- 👎 All providers are imported, including unneeded ones
- 👎 All providers are rendered, including unneeded ones
- 👍 Providers are imported & rendered synchronously
- 👍 It's concise

### Case 2: unified provider, granular rendering

The second proposal, which @tjwiebell has added to this PR, is a configurable unified provider that imports all of the common providers and *conditionally renders* them.

- 👎 All providers are imported, including unneeded ones
- 👍 Only the providers needed are rendered
- 👍 Providers are imported & rendered synchronously
- 👍 It's concise

### Case 3: unified provider, granular importing

A third proposal, which we're discussing, is a configurable unified provider that *conditionally imports* providers and renders them.

- 👍 Only the providers needed are imported
- 👍 Only the providers needed are rendered
- 👎 Providers are imported & rendered asynchronously
- 👍 It's concise

### Non-proposals

To be clear, there is no proposal in which an app would import providers and pass them to the unified provider; that workflow makes zero sense. In such a scenario, the app itself should just render its own providers, since rendering is all the unified provider does.

### Async importing

The point of conflict here is that Case 2 doesn't offer much value over Case 1 because the only cost savings that matter are import cost savings, and Case 2 doesn't offer those savings. Case 3 does, but **async importing providers is not necessarily something we want to do**.

We're envisioning a scenario in which Peregrine contains dozens of providers. If those are dynamically imported, every single consumer of those providers in the app will have to account for a case where the provider isn't present yet, and in some cases will have to duplicate the provider's initial state in order to render meaningfully.

We tried this early on with Redux slices. We figured out how to dynamically import individual reducers and add them to the store on demand, and we were doing it for cart and checkout. Unfortunately, it makes such a logical mess of the code that we stopped doing it, which is why today all reducers are present at first render.

### Other options

We could try categorizing providers as essential and non-essential. In this scenario, the unified provider would import and render the essential providers synchronously, so either Case 1 or Case 2. For non-essential providers, it would be left to the app to import and render them individually, as in Case 0—and it could probably do so asynchronously for certain providers. Thus:

### Case 4: unified provider, for essentials only

- 👍 Only the providers needed are imported, mostly
- 👍 Only the providers needed are rendered
- 👍 Providers are imported & rendered synchronously
- 👍 It's concise